src/services/arweave/ArweaveCompositeDataProvider.ts

```diff
@@ -1,36 +1,46 @@
-import { ArweaveTransactionID, TransactionTag } from '../../types';
 import {
   ANTContractJSON,
   ArNSContractJSON,
   ArweaveDataProvider,
-  SmartweaveDataProvider,
+  ArweaveTransactionID,
+  SmartweaveContractCache,
+  SmartweaveContractInteractionProvider,
+  TransactionTag,
 } from '../../types';
 
 export class ArweaveCompositeDataProvider
-  implements SmartweaveDataProvider, ArweaveDataProvider
+  implements
+    SmartweaveContractInteractionProvider,
+    SmartweaveContractCache,
+    ArweaveDataProvider
 {
   // NOTE: this class should not have any logic for performing queries itself, but rather logic for getting results from
   // an array of providers, using different strategies such as Promise.race or Promise.all.
-  private _warpProvider: SmartweaveDataProvider;
+  private _interactionProvider: SmartweaveContractInteractionProvider;
+  private _contractProviders: SmartweaveContractCache[];
   private _arweaveProvider: ArweaveDataProvider;
 
   // TODO: implement strategy methods
   constructor(
-    warpProviders: SmartweaveDataProvider,
-    arweaveProviders: ArweaveDataProvider,
+    arweaveProvider: ArweaveDataProvider,
+    interactionProvider: SmartweaveContractInteractionProvider,
+    contractProviders: SmartweaveContractCache[],
   ) {
-    this._warpProvider = warpProviders;
-    this._arweaveProvider = arweaveProviders;
+    this._contractProviders = contractProviders;
+    this._interactionProvider = interactionProvider;
+    this._arweaveProvider = arweaveProvider;
   }
 
   async getWalletBalance(id: ArweaveTransactionID): Promise<number> {
     return this._arweaveProvider.getWalletBalance(id);
   }
 
-  async getContractState(
+  async getContractState<T extends ArNSContractJSON | ANTContractJSON>(
     id: ArweaveTransactionID,
-  ): Promise<ArNSContractJSON | ANTContractJSON | undefined> {
-    return this._warpProvider.getContractState(id);
+  ): Promise<T> {
+    return Promise.any(
```

Comment:
why `any` vs `race` on the promise?

Reply:
we want it to fallback to `warp` if we get a bad response from the pdns-service. `any` gives us the first `resolved` promise, whereas `race` returns the first `rejection` or `resolve`https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race#return_value